### PR TITLE
Fix compatibility warning for multiview

### PR DESCRIPTION
**Pull request by z3moon**
Verify the compatibility between a compiled material and the engine's setting only when the engine is set up for stereo.

Materials are always compiled with either 'instanced' or 'multiview even if we compile them with 'none'. Therefore Filament will emit warnings unintentionally in this case. This commit fixes it.

### CODE REVIEW

**Comment by poweifeng** [comment]
> Verify the compatibility between a compiled material and the engine's setting only when the engine is set up for stereo.
> 
> Materials are always compiled with either 'instanced' or 'multiview even if we compile them with 'none'. Therefore Filament will emit warnings unintentionally in this case. This commit fixes it.

but what is the meaning of `VARIANT_HAS_STEREO` in the output shaders?  Is it always true?

**Comment by z3moon** [comment]
> > Verify the compatibility between a compiled material and the engine's setting only when the engine is set up for stereo.
> > Materials are always compiled with either 'instanced' or 'multiview even if we compile them with 'none'. Therefore Filament will emit warnings unintentionally in this case. This commit fixes it.
> 
> but what is the meaning of `VARIANT_HAS_STEREO` in the output shaders? Is it always true?

VARIANT_HAS_STEREO is true for variants including the STE flag. It's likely all materials have some variants including it. But, I think the comments are not accurate yet. Let me update it.

**Comment by poweifeng** [comment]
> > > Verify the compatibility between a compiled material and the engine's setting only when the engine is set up for stereo.
> > > Materials are always compiled with either 'instanced' or 'multiview even if we compile them with 'none'. Therefore Filament will emit warnings unintentionally in this case. This commit fixes it.
> > 
> > 
> > but what is the meaning of `VARIANT_HAS_STEREO` in the output shaders? Is it always true?
> 
> VARIANT_HAS_STEREO is true for variants including the STE flag. It's likely all materials have some variants including it. But, I think the comments are not accurate yet. Let me update it.

Then I think you can replace the material domain check with a variant check logic (e.g `FMaterial::hasVariant()`) for `Variant::STE`?

**Comment by z3moon** [comment]
> > > > Verify the compatibility between a compiled material and the engine's setting only when the engine is set up for stereo.
> > > > Materials are always compiled with either 'instanced' or 'multiview even if we compile them with 'none'. Therefore Filament will emit warnings unintentionally in this case. This commit fixes it.
> > > 
> > > 
> > > but what is the meaning of `VARIANT_HAS_STEREO` in the output shaders? Is it always true?
> > 
> > 
> > VARIANT_HAS_STEREO is true for variants including the STE flag. It's likely all materials have some variants including it. But, I think the comments are not accurate yet. Let me update it.
> 
> Then I think you can replace the material domain check with a variant check logic (e.g `FMaterial::hasVariant()`) for `Variant::STE`?

The surface domain check is needed as we don't care for other domains. Let me add additional check for STE flag.

**Review by pixelflinger — APPROVED** [review]

File: filament/src/details/Material.cpp [hunk 1]
```diff
@@ -192,19 +192,27 @@ Material* Material::Builder::build(Engine& engine) {
         return nullptr;
     }
 
+    // Print a warning if the material's stereo type doesn't align with the engine's setting.
     MaterialDomain materialDomain;
+    UserVariantFilterMask variantFilterMask;
     materialParser->getMaterialDomain(&materialDomain);
-    if (materialDomain == MaterialDomain::SURFACE) {
+    materialParser->getMaterialVariantFilterMask(&variantFilterMask);
+    bool const hasStereoVariants = !(variantFilterMask & UserVariantFilterMask(UserVariantFilterBit::STE));
```
[the remark this answers is not in the document]
Comment (z3moon) [reply filament/src/details/Material.cpp:200]: The overall material impl implies it considers only blacklists. So I doubt it'd change as it's very drastic. If it's changed, all other places should be rewritten including this.

**Pull request merged**